ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs

```diff
@@ -141,9 +171,8 @@ data NetworkNode addr m r = NetworkNode {
       -- will run this to connect with a given list of peer.  But it can also
       -- be used at a later stage to connect to a new peer.
       --
-      -- This function will run client side of mux version negotation and then
-      -- start a the list protocols given by @'NetworkInterface'@.
-      connect :: WithConnection m addr (Connection m) r,
+      -- This function will run client side of mux version negotation and then run client side of @'MuxAplication@' on it.
+      connect :: addr -> (m () -> m r) -> m r,
```

Comment:
Presumably the `m ()` returns when the connection is done?

Or does the `m ()` run the connection?

What happens if I use the `m ()` twice? What happens if I never use it?

Why not parameterize the return type rather than fix to `()`? Would probably be useful in testing/simulation.

A more sensible type IMO would be `addr -> m ()` because, after all, what is _done_ with a connection is fixed by that `MuxApplication` value anyway, right?

Reply:
@avieth I agree, it should be just `addr -> m ()`, and then one can do with that returned `m ()` whatever one want (run it or not or run it twice).  One can always use it twice, in either form :).  And I think it will just run fine - after all this will connect to some upstream peer and communicate with it, e.g.  one'll receive blocks twice.

Reply:
Hmm, I see why I did this in this way.  Because under the hood it handles resources.  This `connect` is something like this: 
`\k -> bracket socket close \sd -> let conn = ... in k conn`
and when `k` exits the socket is closed.  Maybe I should call it `withConnection` instead.

Reply:
Huh, it's still weird. Since the programmer at this scope is not actually controlling anything about the connection, it's bizarre to have that the connection itself is limited to the continuation. I mean, the programmer doesn't even get any return value from the connection.

Reply:
This thing essentially says "do something magic in some other thread and I hope I'll see some effects pop up". A bracketing pattern would make sense it a `Connection` value were given, and _that_ was used to run the mux client there, in this thread of control. As it stands now, it's difficult to grapple: I would tend to believe that the `MuxApplication` alone determines when the connection stops.

Reply:
You're right.  There is one more problem to it: is would keep the socket description as long as the callback was running, and thus longer than needed.  Also the only sensible argument is `id :: a -> a`, so it's better to apply it.